presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java

```diff
@@ -686,6 +749,45 @@ private List<TaskSource> getTaskSources(Iterator<SerializedPrestoSparkTaskSource
         return result.build();
     }
 
+    private List<TaskSource> getNativeExecutionShuffleSources(
+            Session session, TaskId taskId, PlanFragment fragment, Map<PlanNodeId, PrestoSparkShuffleReadInfo> shuffleReadInfos, List<TaskSource> taskSources)
+    {
+        ImmutableSet.Builder<ScheduledSplit> result = ImmutableSet.builder();
+        PlanNode root = fragment.getRoot();
+        AtomicLong nextSplitId = new AtomicLong();
+        shuffleReadInfos.forEach((planNodeId, info) ->
+                result.add(new ScheduledSplit(nextSplitId.getAndIncrement(), planNodeId, new Split(REMOTE_CONNECTOR_ID, new RemoteTransactionHandle(), new RemoteSplit(
+                        new Location(format("batch://%s?shuffleInfo=%s", taskId, shuffleInfoTranslator.createSerializedReadInfo(info))),
+                        taskId)))));
+
+        List<TaskSource> nativeExecutionSources = taskSources.stream().filter(taskSource -> isNativeExecutionNode(taskSource.getPlanNodeId(), root)).collect(Collectors.toList());
+        checkArgument(nativeExecutionSources.size() <= 1, "At most 1 taskSource is expected for NativeExecutionNode but got %s", nativeExecutionSources.size());
```

Comment:
In Java, 

checkArgument is for validating argument.

checkState is for verifying internal constraints, this method looks more like checkState than checkArgument.